Fix use scope and highlighting for declarations (Issue #235)

eliasku opened the pull request:
This PR will fix highlighting issue #235 and will fix renaming from parameter or local var/function declarations.

- Fixed usage scope for variables/functions/parameters declaration psi elements.
- Highlighting from usage to declaration fixed
- Renaming from declaration parameter fixed (test case added)
- Unit test added for renaming which isn't work on previous implementation
- Release notes updated

**Before:**
![image](https://cloud.githubusercontent.com/assets/3038174/13331904/a326d208-dc10-11e5-9e51-456f4731524f.png)
![image](https://cloud.githubusercontent.com/assets/3038174/13331921/b4e98f3a-dc10-11e5-8deb-8d41ab2c8cd8.png)

**After:**
![image](https://cloud.githubusercontent.com/assets/3038174/13331824/547b499a-dc10-11e5-869d-0088da2745c1.png)
![image](https://cloud.githubusercontent.com/assets/3038174/13331837/64241430-dc10-11e5-971a-9a4883207748.png)

EBatTiVo commented on src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java:
```diff
@@ -62,25 +64,16 @@ public String getName() {
-    return new HaxeComponentNameImpl(getNode());
-    ////this.getName();
-    //return findNotNullChildByClass(HaxeComponentName.class);
+    return null;
```
Why this change?

eliasku replied:
There was NullPointerException from created HaxeComponentNameImpl, because the passed AST-Node had wrong structure sometimes.

EBatTiVo replied:
So, the answer is to strip the implementation?  The function is obviously used (or you wouldn't have gotten the NPE), so some other functionality depends upon it.

EBatTiVo replied:
There are 41 usages of this function.  I think we still need an implementation.  Was it getNode() that was returning a bad (or unexpected) node?  When/where did the NPE show up?

eliasku replied:
getNode() was right, and NameComponentImpl was created, and then NameComponentImpl::getIdentifier - `findNotNullChildByClass(HaxeIdentifier.class)` throws exception because of `findNotNull__` guarding. I don't remember the case exactly. I could spend a little amount of time to reproduce this case.

NPE had beed shown up for VarDeclaration, not VarPartDeclaration. Notice that getComponentName is overridden for all Implementations, except of VarDeclaration(part is overridden). So we've guarded `getNameIdentifier` which will be Null for VarDeclaration (because the component name inside VarPart).

eliasku replied:
Now, I'm searching the way to reproduce NPE in previous implementation.

EBatTiVo replied:
How about something like this:

PsiIdentifier id = getNameIdentifier();
return null != id ? new HaxeComponenetNameImp(id.getNode()) : null;

eliasku replied:
To reproduce - just type `var a = 0;` and move caret back `var a = <caret>0;` and for LocalVarDeclaration getIdentifier will be required from getComponentName.
```
PsiIdentifier id = getNameIdentifier();
return null != id ? new HaxeComponenetNameImp(id.getNode()) : null;
```
Will cause endless recursion. Notice that `getNameIdentifier` is declared for Mixin

EBatTiVo replied:
> Will cause endless recursion. 

Using your new code, yes.  I was browsing my sources, not the change list, so I got confused.  However, now that the getNameIdentifier() calls getComponentName(), and getComponentName() always returns null, we've stripped the implementation out of *two* functions.

> Notice that getNameIdentifier is declared for Mixin

I don't see a mixin class anywhere in the superclass hierarchy.  Are you saying that *all* subclasses implement it?  If we are going to depend upon that, then we should just remove the method from this (abstract) class; then we'll enforce it.

EBatTiVo commented on src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java:
```diff
@@ -213,4 +206,19 @@ public boolean hasModifierProperty(@HaxePsiModifier.ModifierConstant @NonNls @No
+    if(parent != null) {
+      if (this instanceof HaxeLocalVarDeclarationPart) {
+        // parent of local var declaration-Part
+        return new LocalSearchScope(parent.getParent());
```
I'm growing concerned about code like this.  If we have to expand the tree for some reason, then code that expects a parent of a given type at a certain place, then code like this will break.  And I know I've been taking (and probably writing) code like this a lot.  I think that perhaps we should be using UsefulPsiTreeUtil.getParentOfType() instead of expecting a specific element's tree position.  

What do you think?

eliasku replied:
Agreed. Fixed.

And code with manual children traversing could be broken for new psi structure (like we saw in getUnderlyingType method before). I will focus more attention to it by the next coding and reviewing

eliasku commented:
1. I've reverted `getComponentName` to your suggestion to create `NameComponentImpl` in case if `getNameIdentifier` returns not-null. (Purpose is to **Fix NPE**)

2. I've fixed the `getNameIdentifier` from previous logic to new one. Previous logic was breaking usages to local var declaration referencing. (Purpose is to **Fix highlighting issue**)

3. I've made typed searching. Block-statement type parent for LocalSearchScope creation, but only in case if we're in context of Local-Declaration-Node. (I'm not sure 100% how IDEA works with `getUseScope`, because previously I thought that IDEA will not search in comments outside of LocalSearchScope while refactoring. It was wrong. But at least it fixes the renaming of same names  references outside )

EBatTiVo commented on src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java:
```diff
@@ -213,4 +208,16 @@ public boolean hasModifierProperty(@HaxePsiModifier.ModifierConstant @NonNls @No
+  public SearchScope getUseScope() {
+    final PsiElement localVar = UsefulPsiTreeUtil.getParentOfType(this, HaxeLocalVarDeclaration.class);
+    if(localVar != null) {
+      final PsiElement outerBlock = UsefulPsiTreeUtil.getParentOfType(localVar, HaxeBlockStatement.class);
```
So, does it work in this case?:

```
class bar {
  function doSomething();
}
class foo {
  var a:bar = new bar; 
  function printNtimes(msg:String, n:Int) :Void for( a in 0...n ) Sys.println( a.<caret>_toString() + ":" + msg);
  //                                                                              ^^^
}
```

eliasku replied:
It doesn't work because `a` inside `for` is not local-variable-declaration. For `for` statement I've added support for renaming from using to declaration. It was missing NameIdentifier. Look new commit, your example will work correctly.

EBatTiVo replied:
That is a good fix.  However, the point of the sample was really to point out that local scopes (such as functions) aren't *always* surrounded by HaxeBlockStatements.  In my version of the example, the variable 'a' inside of the for loop hides the variable with the same name and different type declared at the class level.  Therefore, locating the scope requires searching for more than an enclosing block statement; it requires looking for enclosing function definitions, too.

EBatTiVo replied:
I guess I missed this statement you made.

> It doesn't work because a inside for is not local-variable-declaration.

If a variable being declared on a for statement is not a local variable declaration, what is it?  More importantly: Would somebody writing code expect searching, highlighting, refactoring, etc. to act as if it *were* a local variable?

EBatTiVo replied:
And then, after making these arguments, I go look at your unit test more closely and it's working as I expect.

eliasku commented:
Thank you, @EBatTiVo! `for` statement declaration is also fixed now!
Before:
![image](https://cloud.githubusercontent.com/assets/3038174/13393991/0e10e7c4-def6-11e5-9465-cff29afb0f85.png)
After:
![image](https://cloud.githubusercontent.com/assets/3038174/13394034/5cb00f86-def6-11e5-9924-d8e2e75ad523.png)

EBatTiVo commented:
@eliasku Would you mind looking at #136, which is very closely related, and see if this change fixed that, too?

eliasku commented:
@EBatTiVo I think so. I reproduce this issue #136 it in `0.9.9`. For actual `origin/master` there is no issue anymore.

EBatTiVo commented:
That wording is a bit confusing.  I take it to mean that you tested and the issue is fixed.  Correct?

eliasku commented:
That's right. I've corrected the previous comment.
